cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesDataset.java

```diff
@@ -0,0 +1,177 @@
+      default:
+        return new HashMap<>();
+    }
+  }
+
+  public Map<String, String> getResolvedProperties(EntityId entityId) {
+    // if it is instance level get the properties and return
+    if (entityId.getEntityType().equals(EntityType.INSTANCE)) {
+      return getProperties(entityId);
+    }
+
+    // get properties from current level
+    Map<String, String> properties = getProperties(entityId);
+    // if the entity id has a parent id, get the preference from its parent
+    if (entityId instanceof ParentedId) {
+      properties.putAll(getProperties(((ParentedId) entityId).getParent()));
+    } else {
+      // otherwise it is a namespace id, which we want to look at the instance level
+      properties.putAll(getProperties(new InstanceId("")));
+    }
+    return properties;
+  }
+
+  public void setProperties(EntityId entityId, Map<String, String> propMap) {
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        setConfig(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES), propMap);
+        break;
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        setConfig(namespaceId.getNamespace(), getMultipartKey(namespaceId.getNamespace()), propMap);
+        break;
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        setConfig(appId.getNamespace(), getMultipartKey(appId.getNamespace(), appId.getApplication()), propMap);
+        break;
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        setConfig(programId.getNamespace(),
+                  getMultipartKey(programId.getNamespace(), programId.getApplication(),
+                                  programId.getType().getCategoryName(), programId.getProgram()), propMap);
+        break;
+      default:
+        // do nothing
+    }
+  }
+
+  public void deleteProperties(EntityId entityId) {
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        deleteConfig(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES));
+        break;
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        deleteConfig(namespaceId.getNamespace(), getMultipartKey(namespaceId.getNamespace()));
+        break;
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        deleteConfig(appId.getNamespace(), getMultipartKey(appId.getNamespace(), appId.getApplication()));
+        break;
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        deleteConfig(programId.getNamespace(),
+                     getMultipartKey(programId.getNamespace(), programId.getApplication(),
+                                     programId.getType().getCategoryName(), programId.getProgram()));
+        break;
+      default:
+        // do nothing
+    }
+  }
+
+  private String getMultipartKey(String... parts) {
+    int sizeOfParts = 0;
```

Comment:
You can use Stream instead

`int sizeOfParts = Stream.of(parts).mapToInt(String::length).reduce(0, (a, b) -> a + b)`

Reply:
Fixed